src/main/java/org/broadinstitute/hellbender/tools/spark/sv/InsertSizeMetricsCollectorSpark.java

```diff
@@ -1,99 +1,256 @@
+     * @param accumLevels           accumulation level {ALL_READS, SAMPLE, LIBRARY, READ_GROUP}
+     * @param histogramMADTolerance MAD tolerance when producing histogram plot
      */
-    // TODO: * @param accumLevels            accumulation level {SAMPLE, LIBRARY, READGROUP}
     public InsertSizeMetricsCollectorSpark(final JavaRDD<GATKRead> filteredReads,
-                                           //final Set<MetricAccumulationLevel> accumLevels,
-                                           final List<SAMReadGroupRecord> samRgRecords,
-                                           final double HistogramMADTolerance) {
-
-        final JavaRDD<Integer> lengthRDD = filteredReads.map(read -> Math.abs(read.getFragmentLength()) );
-        lengthRDD.cache();
-
-        metrics.READ_PAIRS = filteredReads.count();
-
-        collectLevelInfo();
+                                           final SAMFileHeader header,
+                                           final Set<MetricAccumulationLevel> accumLevels,
+                                           final double histogramMADTolerance) {
+
+        // construct untrimmed hand rolled "histogram" (SortedMap) in three steps, because htsjdk Histogram does not play well with Serialization
+        // so first hand roll a histogram using Spark (read traversal is the bottleneck in terms of performance),
+        //   when computing corresponding metrics locally, use htsjdk Histogram converted the sorted map version to do the work
+        final Map<GroupInfo, Map<SamPairUtil.PairOrientation, SortedMap<Integer, Long>>> unsortedHistogramsAtRGLevel = filteredReads.mapToPair(read -> step1ReadTraversal(read, header))
+                                                                                                                                    .groupByKey()
+                                                                                                                                    .mapToPair(InsertSizeMetricsCollectorSpark::step2ConvertToList)
+                                                                                                                                    .mapToPair(InsertSizeMetricsCollectorSpark::step3ConvertToHistogram)
+                                                                                                                                    .collectAsMap();
+
+        final SortedMap<GroupInfo, Map<SamPairUtil.PairOrientation, SortedMap<Integer, Long>>> histogramsAtRGLevel = new TreeMap<>(groupInfoComparator);
+        histogramsAtRGLevel.putAll(unsortedHistogramsAtRGLevel);
+
+        LinkedList<SortedMap<GroupInfo, Map<SamPairUtil.PairOrientation, SortedMap<Integer, Long>>>> listOfStats = new LinkedList<>();
+
+        if (accumLevels.contains(MetricAccumulationLevel.READ_GROUP)) {
+            listOfStats.addFirst(histogramsAtRGLevel);
+        }
+        if (accumLevels.contains(MetricAccumulationLevel.LIBRARY)) {
```

Comment:
This approach will take several cycles over the read group stats lists, and it seems like it would be possible to do this all in one pass. It's probably not a big deal performance-wise, so not a dealbreaker, but think about it.